### PR TITLE
add test explorer

### PR DESCRIPTION
**Pull request by ansyral**
Signed-off-by: xuzho <xuzho@microsoft.com>

right now supports:

1. sync test explorer when test updates
2. click on test explorer, jump to editor selection of the test

[to-do]: trigger tests from test explorer.

it now looks like below:
image

### CODE REVIEW

File: com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/internal/JUnitTestSearcher.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+				IJavaSearchConstants.ANNOTATION_TYPE_REFERENCE,
+				SearchPattern.R_EXACT_MATCH | SearchPattern.R_CASE_SENSITIVE);
+		SearchPattern pattern = SearchPattern.createOrPattern(runWithPattern, testPattern);
+		List<TestSuite> tests = new ArrayList<>();
+		HashSet<IType> testClasses = new HashSet<>();
+
+		SearchRequestor requestor = new SearchRequestor() {
+			@Override
+			public void acceptSearchMatch( SearchMatch match ) throws CoreException {
+
```
Comment (yaohaizh) [inline com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/internal/JUnitTestSearcher.java]: Fomatting?

File: com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/internal/JUnitTestSearcher.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+					uri,
+					type.getFullyQualifiedName()  + "#" + member.getElementName(),
+					type.getPackageFragment().getElementName(),
+					TestLevel.Method);
+		}
+	}
+	
+	private Range getRange(ICompilationUnit typeRoot, IJavaElement element) throws JavaModelException {
+		ISourceRange r = ((ISourceReference) element).getNameRange();
+		final Range range = JDTUtils.toRange(typeRoot, r.getOffset(), r.getLength());
```
Comment (yaohaizh) [inline com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/internal/JUnitTestSearcher.java]: Just return value?

File: com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/internal/JUnitTestSearcher.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+			        testClasses.add(type);
+				}
+			}
+		};
+
+		try {
+			ITypeHierarchy hierarchy = JavaCore.newTypeHierarchy(region, null, null);
+		    IType[] allClasses = hierarchy.getAllClasses();
+		    IJavaSearchScope scope =
+		            SearchEngine.createJavaSearchScope(allClasses, IJavaSearchScope.SOURCES);
```
Comment (yaohaizh) [inline com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/internal/JUnitTestSearcher.java]: If we can constrain the search scope to the project source files, it doesn't need the region any more.

File: extension/src/testResourceManager.ts [hunk 4]
```diff
@@ -34,6 +39,35 @@ export class TestResourceManager {
         const test = this.getTests(file);
         return test ? test.dirty : undefined;
     }
+    public getAll(): TestSuite[] {
+        let allTests: TestSuite[] = [];
+        this.testsIndexedByFileUri.forEach((value, key, m) => {
+            allTests = [...allTests, ...value.tests];
```
Comment (yaohaizh) [inline extension/src/testResourceManager.ts]: Arrays.concat is okay for this case

File: extension/src/Explorer/testExplorer.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            case TestTreeNodeLevel.Package:
+                return (_) => _.packageName;
+            case TestTreeNodeLevel.Class:
+                return (_) => this.getShortName(_.parent);
+            default:
+                throw new Error('Not supported group level');
+        }
+    }
+
+    private getWkspaceFolder(test: TestSuite): string {
```
Comment (yaohaizh) [inline extension/src/Explorer/testExplorer.ts]: getWorkspaceFolder?

**Comment by ansyral** [comment]
#45

File: extension/src/extension.ts [hunk 6]
```diff
@@ -44,24 +46,38 @@ let running: boolean = false;
...
     workspace.onDidChangeTextDocument((event) => {
         const uri = event.document.uri;
         testResourceManager.setDirty(uri);
         // onDidChange.fire();
     });
 
+    workspace.onDidSaveTextDocument((document) => {
+        const uri = document.uri;
+        testResourceManager.setDirty(uri);
+        onDidChange.fire();
```
Comment (akaroml) [inline extension/src/extension.ts]: Who is listening to this event?

Comment (ansyral) [reply extension/src/extension.ts]: junitCodelensProvider

File: extension/src/Explorer/testExplorer.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+import { window, workspace, Event, EventEmitter, ExtensionContext, TreeDataProvider, TreeItem, TreeItemCollapsibleState, Uri, ViewColumn, Command } from "vscode";
+import * as Commands from '../commands';
+import { TestLevel, TestSuite } from '../protocols';
+import { TestResourceManager } from '../testResourceManager';
+import { TestTreeNode, TestTreeNodeLevel } from "./testTreeNode";
+
+export class TestExplorer implements TreeDataProvider<TestTreeNode> {
+    private _onDidChangeTreeData: EventEmitter<TestTreeNode | undefined> = new EventEmitter<TestTreeNode | undefined>();
+    // tslint:disable-next-line
+    public readonly onDidChangeTreeData: Event<TestTreeNode | null> = this._onDidChangeTreeData.event;
```
[the remark this answers is not in the document]
Comment (ansyral) [reply extension/src/Explorer/testExplorer.ts]: currently no

File: extension/src/extension.ts (hunk 6 above)
Comment (ansyral) [reply extension/src/extension.ts]: found that sometimes when editing the test code, codelens wouldn't update. So i add this to ensure codelens are refreshed by sure. see #36

File: extension/src/Explorer/testTreeNode.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import { Range } from "vscode";
+
+export class TestTreeNode {
+    constructor(
+        private _name: string,
```
[the remark this answers is not in the document]
Comment (ansyral) [reply extension/src/Explorer/testTreeNode.ts]: in `c#`, we prefer property to public field as it provide better encapsulation. For `typescript`, i'm not sure whether it is the same way. personally, i think property is safer

**Comment by ansyral** [comment]
new look:
image

**Pull request merged**